optuna/_hypervolume/box_decomposition.py

```diff
@@ -0,0 +1,144 @@
+
+    upper_bound_set = np.asarray([ref_point])  # Line 1 of Alg. 2.
+    def_points = np.full((1, n_objectives, n_objectives), -np.inf)  # z^k(z^r) = \hat{z}^k
+    def_points[0, objective_indices, objective_indices] = ref_point  # \hat{z}^k is a dummy point.
+    for solution in sorted_pareto_sols:  # NOTE(nabenabe): Sorted must be fulfilled.
+        upper_bound_set, def_points = update(solution, upper_bound_set, def_points)
+
+    return upper_bound_set, def_points
+
+
+def _get_box_bounds(
+    def_points: np.ndarray, upper_bound_set: np.ndarray, ref_point: np.ndarray
+) -> np.ndarray:
+    # Eq. (2) of Lacour17.
+    n_objectives = upper_bound_set.shape[-1]
+    bounds = np.empty((2, *upper_bound_set.shape))
+    bounds[0, :, 0] = def_points[:, 0, 0]
+    bounds[1, :, 0] = ref_point[0]
+    row, col = np.diag_indices(n_objectives - 1)
+    bounds[0, :, 1:] = np.maximum.accumulate(def_points, axis=-2)[:, row, col + 1]
+    bounds[1, :, 1:] = upper_bound_set[:, 1:]
+    not_empty = ~np.any(bounds[1] <= bounds[0], axis=-1)  # Remove [inf, inf] or [-inf, -inf].
+    return bounds[:, not_empty]
+
+
+def _get_non_dominated_box_bounds(
+    sorted_pareto_sols: np.ndarray, ref_point: np.ndarray
+) -> tuple[np.ndarray, np.ndarray]:  # (n_bounds, n_objectives) and (n_bounds, n_objectives)
+    # The calculation of u[k] and l[k] in the paper: https://arxiv.org/abs/2006.05078
+    # NOTE(nabenabe): The paper handles maximization problems, but we consider minimization here.
+    _upper_bound_set = _get_upper_bound_set(sorted_pareto_sols, ref_point)[0]
+    upper_bound_set = _upper_bound_set[np.argsort(_upper_bound_set[:, 0])[::-1]]
+    # Flip the sign and use upper_bound_set as the Pareto solutions. Then we can calculate the
+    # lower bound set as well.
+    point_at_infinity = np.full_like(ref_point, np.inf)
+    neg_upper_bound_set, neg_def_points = _get_upper_bound_set(-upper_bound_set, point_at_infinity)
+    ubs, lbs = -_get_box_bounds(neg_def_points, neg_upper_bound_set, point_at_infinity)
+    return lbs, ubs
+
+
+def get_non_dominated_box_bounds(
+    loss_vals: np.ndarray, ref_point: np.ndarray
+) -> tuple[np.ndarray, np.ndarray]:  # (n_bounds, n_objectives) and (n_bounds, n_objectives)
+    assert np.all(np.isfinite(loss_vals)), "loss_vals must be clipped before box decomposition."
+    # Remove duplications and lexsort the solutions by ``np.unique``.
+    unique_lexsorted_loss_vals = np.unique(loss_vals, axis=0)
+    sorted_pareto_sols = unique_lexsorted_loss_vals[
+        _is_pareto_front(unique_lexsorted_loss_vals, assume_unique_lexsorted=True)
+    ]
+    n_objectives = loss_vals.shape[-1]
+    # The condition here follows BoTorch.
+    # https://github.com/pytorch/botorch/blob/v0.13.0/botorch/acquisition/multi_objective/utils.py#L55-L63
+    if n_objectives > 4:
+        warnings.warn(
+            "GPSampler might be significantly slow for n_objectives > 4. "
+            "Please consider using another sampler instead."
```

Comment:
It would be great not to mention `GPSampler` here since this `get_non_dominated_box_bounds` is independent from it. How about the following?
```suggestion
            "Box decomposition of solutions might be significantly slow for n_objectives > 4. "
            "Please consider using another sampler instead."
```

Reply:
Ah, wait, we say `Please consider using another sampler instead`, but should we also change this part as well?
Because the first sentence talks about box decomposition and the second sentence talks about sampler, which could be a bit confusing for users?

Reply:
Sounds good to change the second sentence.

Reply:
I thought the same thing!